### PR TITLE
test: add test for re_thread_enter() and leave()

### PR DESCRIPTION
**Pull request by alfredh**
Add testcase using mqueue for:

* re_thread_enter()
* re_thread_leave()

### CODE REVIEW

**Comment by alfredh** [comment]
seems like it is not working properly:

```
multithread: 185 tests with 2 repeats (total 370 threads): main: thread check: called from a NON-RE thread without thread_enter()!
main: thread check: called from a NON-RE thread without thread_enter()!
main: thread check: called from a NON-RE thread without thread_enter()!
main: thread check: called from a NON-RE thread without thread_enter()!
```

**Comment by sreimers** [comment]
If I remember correctly, `re_thread_enter` and `re_thread_leave` works only with one re poll thread (`re_global`). otherwise it does not now which re poll thread to enter, since it's not specified.

You can use:

```c
	if (test_mode == TEST_THREAD)
		return ESKIPPED;
```

**Comment by alfredh** [comment]
```
(gdb) r
Starting program: /home/alfredh/git/re/test/retest -v -r test_remain
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
single testcase: test_remain
retest: libre version 3.4.0 (x86_64/Linux)
using async polling method 'epoll'
using datapath './test/data'
regular tests:       [New Thread 0x7ffff773f6c0 (LWP 222302)]
main: thread_init: already added for thread
[Thread 0x7ffff773f6c0 (LWP 222302) exited]
[New Thread 0x7ffff773f6c0 (LWP 222303)]
remain: thread start
main: thread check: called from a NON-RE thread without thread_enter()! [te=0, eq=0]
.... tid=140737352963520
.... cur=140737344960192
main: thread check: called from a NON-RE thread without thread_enter()! [te=0, eq=0]
.... tid=140737352963520
.... cur=140737344960192
remain: starting re_main loop..
[Thread 0x7ffff773f6c0 (LWP 222303) exited]
```

**Comment by alfredh** [comment]
the thread enter check is unreliable:

```
single testcase: test_remain
retest: libre version 3.4.0 (x86_64/Linux)
using async polling method 'epoll'
using datapath './test/data'
regular tests:       remain: thread start
main: re_thread_enter
main: re_thread_enter: Not Polling
remain: starting re_main loop..
main: thread check: called from a NON-RE thread without thread_enter()! [thread_enter=0]
.... tid=139827025360320
.... cur=139827017414336
re main loop:
  maxfds:       1024
  nfds:         0
  method:       epoll
  polling:      1
  sig:          0
  timers:       1
  mutex:        0x558f92081e80
  tid:          0x7f2c042d35c0
  thread_enter: 0
  async:        (nil)

main: thread check: called from a NON-RE thread without thread_enter()! [thread_enter=0]
.... tid=139827025360320
.... cur=139827017414336
re main loop:
  maxfds:       1024
  nfds:         0
  method:       epoll
  polling:      1
  sig:          0
  timers:       1
  mutex:        0x558f92081e80
  tid:          0x7f2c042d35c0
  thread_enter: 0
  async:        (nil)

main: re_thread_leave: setting THREAD_ENTER to false
```

**Comment by sreimers** [comment]
Yes this is a race condition, since polling loop is running after `thread_enter` it's not locked before. I think we can remove the active polling check within `re_thread_enter`. The intention was if no poll loop is running there is not need for locking, I think.

**Comment by alfredh** [comment]
Thanks, removing the re.polling check made the test stable.

I would prefer that all tests are ultra stable and rock solid, with 100% predictable result.

File: test/remain.c [hunk 1]
```diff
@@ -138,11 +140,81 @@ static int test_remain_thread(void)
...
+
+static int enterleave_thread_handler(void *arg)
+{
+	struct data *data = arg;
+
+	/* Enter an 're' thread */
+	re_thread_enter();
+
+	/* note: allocated from this thread */
+	int err = mqueue_alloc(&data->mq, mqueue_handler, data);
```
Comment (sreimers) [inline test/remain.c:164]: If this returns ENOMEM in OOM memory test `re_cancel` is not called and re_poll timeouts, I think.

Comment (sreimers) [reply test/remain.c:164]: Maybe starting a fallback timer could help. Or ignoring `re_main_timeout` error.

**Comment by alfredh** [comment]
Getting some random errors:

```
remain: TEST_ERR: /home/runner/work/re/re/test/remain.c:188: (Connection timed out [110])

remain: TEST_ERR: /home/runner/work/re/re/test/remain.c:215: (Connection timed out [110])
test: test_remain: test failed (Connection timed out [110])

retest: TEST_ERR: /home/runner/work/re/re/test/main.c:238: (Connection timed out [110])
```

**Pull request closed without merging**